### PR TITLE
Test each benchmark case in own #[test]

### LINKED ISSUE #9110: improve impl_benchmark_test_suite so that it generate a test for each benchmarks
It is currently very hard to debug individual test because it run all the tests and `assert!(!anything_failed);`

### PR DESCRIPTION
**Pull request by ucover**
Updates the `impl_benchmark_test_suite!` and extends the `benchmarks!` macros.  
Closes #9110

### Syntax extension
It is now possible to use `impl_benchmark_test_suite!` **within** the `benchmarks!` macro.  

Example:  
```rust
benchmarks! {
	dummy {
		…
	}: {
		…
	}

	impl_benchmark_test_suite!(Pallet, crate::tests::new_test_ext(), crate::tests::Test) // ; allowed here
}

```
### Behavioural change
When `impl_benchmark_test_suite!` is used in a `benchmarks!` call, it now generates one `#[test]` function for each bench case.

Old output:   
```pre
test benchmarking::benchmark_tests::test_benchmarks ... ok
```
New output:
```pre
test benchmarking::bench_sort_vector ... ok
test benchmarking::bench_accumulate_dummy ... ok
test benchmarking::bench_set_dummy_benchmark ... ok
```

👉 `benchmarks!` and `impl_benchmark_test_suite!` stay backwards compatible.

### Parallel benchmarks

Since each bench case has its own `#[test]` function, it is possible to run them in parallel.  
A good example is the *contracts* pallet:  

Result with old syntax (after compilation):
```sh
cargo test --features=runtime-benchmarks -- --test-threads 16  58,85s user 0,61s system 108% cpu 54,619 total
```

Result with new syntax (after compilation):
```sh
cargo test --features=runtime-benchmarks -- --test-threads 16  65,28s user 0,87s system 264% cpu 24,981 total
```

### Open points

- [x] I am not sure if I ran all test. Some down-stream tests would be good for the compatibility check.
- [ ] Comments need to be checked for changes.
- [ ] I have not bumped the version.

✄ -----------------------------------------------------------------------------

Thank you for your Pull Request! 🙏

Before you submit, please check that:

- [x] **Description:** You added a brief description of the PR, e.g.:
  - What does it do?
  - What important points should reviewers know?
  - Is there something left for follow-up PRs?
- [x] **Labels:** You labeled the PR appropriately if you have permissions to do so:
  - [ ] `A*` for PR status (**one required**)
  - [ ] `B*` for changelog (**one required**)
  - [ ] `C*` for release notes (**exactly one required**)
  - [ ] `D*` for various implications/requirements
  - [ ] Github project assignment
- [x] **Related Issues:** You mentioned a related issue if this PR is related to it, e.g. `Fixes #228` or `Related #1337`.
- [x] **2 Reviewers:** You asked at least two reviewers to review. If you aren't sure, start with GH suggestions.
- [x] **Style Guide:** Your PR adheres to [the style guide](https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md)
  - In particular, mind the maximal line length of 100 (120 in exceptional circumstances).
  - There is no commented code checked in unless necessary.
  - Any panickers in the runtime have a proof or were removed.
- [ ] **Runtime Version:** You bumped the runtime version if there are breaking changes in the **runtime**.
- [ ] **Docs:** You updated any rustdocs which may need to change.

Refer to [the contributing guide](https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.adoc) for details.

✄ -----------------------------------------------------------------------------

### CODE REVIEW

File: frame/balances/src/benchmarking.rs [hunk 1]
```diff
@@ -33,7 +33,12 @@ const SEED: u32 = 0;
 // existential deposit multiplier
 const ED_MULTIPLIER: u32 = 10;
 
-benchmarks_instance_pallet! {
+impl_benchmark_test_suite!(
+	Balances,
+	crate::tests_composite::ExtBuilder::default().build(),
+	crate::tests_composite::Test,
+	by_instance_pallet,
+{
```
Comment (shawntabrizi) [inline frame/balances/src/benchmarking.rs:41]: Not a big fan of this specific syntax. I would prefer we keep the `benchmarks_instance_pallet!` top level macro, and instead have the change required by end users simply be to include the `impl_benchmark_test_suite` code block into the macro body.

does that make sense?

As in most people just need to move the bottom bracket to surround the old syntax, and indent it once.

Comment (shawntabrizi) [reply frame/balances/src/benchmarking.rs:41]:
```
benchmarks_instance_pallet! {
    benchmark_1 { }: _() verify {}
    benchmark_2 { }: _() verify {}
    
    impl_benchmark_test_suite!(
	    Balances,
	    crate::tests_composite::ExtBuilder::default().build(),
	    crate::tests_composite::Test,
    );
}

Comment (ucover) [reply frame/balances/src/benchmarking.rs:41]: >I would prefer we keep the benchmarks_instance_pallet! top level macro, and instead have the change required by end users simply be to include the impl_benchmark_test_suite code block into the macro body.

So this would then be a breaking change, since the `impl_benchmark_test_suite!` macro cannot be used on its own anymore?  
@shawntabrizi

Comment (shawntabrizi) [reply frame/balances/src/benchmarking.rs:41]: @ucover This would not be a breaking change because your macro would simply parse this text, it wouldnt be a real macro.

Comment (shawntabrizi) [reply frame/balances/src/benchmarking.rs]: @thiolliere are there any tricks around that problem? ^

File: frame/benchmarking/src/lib.rs [hunk 2]
```diff
@@ -408,8 +474,40 @@ macro_rules! benchmarks_iter {
...
+			( $( $names )* )
+			( $( $names_extra )* )
+			( $( $names_skip_meta )* )
+			$bench_module,
+			$new_test_ext,
+			$test
+			$(, $( $args )* )?
+		);
+	};
+	// iteration-exit arm which generates one #[test] function for all cases.
```
Comment (thiolliere) [inline frame/benchmarking/src/lib.rs:508]: comment is a bit misleading, in this branch no test are generated. and in both branch the `test_bench_by_name` is implemented (inside impl_benchmark)
```suggestion
	// iteration-exit arm which doesn't generates a #[test] function for all cases.
```

File: frame/benchmarking/src/lib.rs [hunk 3]
```diff
@@ -193,6 +193,7 @@ macro_rules! benchmarks {
 		$( $rest:tt )*
```
Comment (thiolliere) [inline frame/benchmarking/src/lib.rs:193]: I think the doc should be improved with the new functionality.

Also the manual tests should probably be removed, and `extra` and `skip_meta` attribute should be explained but this is not part of this PR.

File: frame/benchmarking/src/lib.rs [hunk 4]
```diff
@@ -695,6 +799,110 @@ macro_rules! benchmark_backend {
...
+					// Skip all #[extra] benchmarks if $extra is false.
+					if !($extra) {
+						let disabled = $crate::vec![ $( stringify!($names_extra).as_ref() ),* ];
+						if disabled.contains(&stringify!($name)) {
+							$crate::log::error!(
+								"INFO: extra benchmark skipped - {}",
+								stringify!($name),
+							);
+							return ();
+						}
```
Comment (thiolliere) [inline frame/benchmarking/src/lib.rs:843]: I don't understand this, disabled is all the extra benchmarks, so what happen for the test for the extra benchmark ? does it always returns early ?

**Review by shawntabrizi — APPROVED** [review]
thank you, looks good to me besides the comments above

File: frame/benchmarking/src/lib.rs [hunk 5]
```diff
@@ -695,6 +799,110 @@ macro_rules! benchmark_backend {
...
+			fn [<bench_ $name>] () {
+				$new_test_exec.$exec_name(|| {
+					// Skip all #[extra] benchmarks if $extra is false.
+					if !($extra) {
+						let disabled = $crate::vec![ $( stringify!($names_extra).as_ref() ),* ];
+						if disabled.contains(&stringify!($name)) {
+							$crate::log::error!(
+								"INFO: extra benchmark skipped - {}",
+								stringify!($name),
+							);
```
Comment (thiolliere) [inline frame/benchmarking/src/lib.rs:841]: why is it an error to skip extra ?

Comment (thiolliere) [inline frame/benchmarking/src/lib.rs:841]: it is a bit weird to log an error for an info
```suggestion
							$crate::log::info!(
								"extra benchmark skipped - {}",
								stringify!($name),
							);
```

Comment (shawntabrizi) [reply frame/benchmarking/src/lib.rs:841]: I think this is so the log appears when running the tests by default.

File: frame/benchmarking/src/lib.rs [hunk 6]
```diff
@@ -695,6 +799,110 @@ macro_rules! benchmark_backend {
...
+									// This is considered a success condition.
+									$crate::log::error!(
+										"WARNING: benchmark error skipped - {}",
+										stringify!($name),
+									);
+								}
+							}
+						},
+						Ok(Ok(())) => (),
+					}
```
Comment (thiolliere) [inline frame/benchmarking/src/lib.rs:886]: I don't think we need to do this complex stuff, and I think this complex gives not good error message, which was the point of the PR. Instead we should just makes panic being regular panic in test, and we should give the error when the benchmark return an error.
So something like this
```suggestion
					// Same per-case logic as when all cases are run in the
					// same function.
					match $module::<$test>::[< test_benchmark_ $name >] () {
						Err(err) => {
							match err {
								$crate::BenchmarkError::Stop(err) => panic!("benchmark return stop: {}", err),
								// This is still considered a success condition.
								$crate::BenchmarkError::Override(_) => (),
								// This is still considered a success condition.
								$crate::BenchmarkError::Skip => (),
							}
						},
						Ok(()) => (),
					}
```

File: frame/benchmarking/src/lib.rs (hunk 5 above)
Comment (thiolliere) [reply frame/benchmarking/src/lib.rs:841]: I feel like if somebody writes `extra = false` then he expects the extra benchmarks to not run. So logging something doesn't look very useful.
Actually I don't know why somebody would write `extra = false` to be honest.

File: frame/benchmarking/src/lib.rs (hunk 6 above)
Comment (thiolliere) [reply frame/benchmarking/src/lib.rs:886]: also here I don't see why we log a warning where it is a correct implementation of a benchmark.

Comment (shawntabrizi) [reply frame/benchmarking/src/lib.rs:886]: The problem here is that you may want to program in a a custom `BenchmarkError` for using the benchmark in production, but not want to trigger that pathway when running a test.

The problem is that not all benchmarks apply to all possible runtime configurations (specifically in the XCM world), which is why i wanted to funnel this data to the top

File: frame/benchmarking/src/lib.rs (hunk 5 above)
Comment (shawntabrizi) [reply frame/benchmarking/src/lib.rs:841]: Yeah I guess only if one of the extra benchmarks is so insane heavy... but yeah, doesnt really make sense to have `extra = false` ever in tests. That is the point of tests.

File: frame/benchmarking/src/lib.rs (hunk 6 above)
Comment (thiolliere) [reply frame/benchmarking/src/lib.rs:886]: Ok my suggestion is wrong, we can keep the log. But moving the println into the panic message gives probably a better error message.

File: frame/benchmarking/src/lib.rs (hunk 5 above)
Comment (thiolliere) [reply frame/benchmarking/src/lib.rs:841]: I see, then ok

**Review by thiolliere — COMMENTED** [review]
looks good to me once doc about new feature is added

File: frame/balances/src/benchmarking.rs (hunk 1 above)
Comment (ucover) [reply frame/balances/src/benchmarking.rs]: I tried it with ` $(;)?` but that did not work. Yours works :thinking:

File: frame/benchmarking/src/lib.rs (hunk 6 above)
Comment (ucover) [reply frame/benchmarking/src/lib.rs:886]: I felt a bit weird for writing that `assert!(false)` :laughing:.
Replaced it with and the `println` with `panic!("{}: {:?}", stringify!($name), err);`.

File: frame/benchmarking/src/lib.rs (hunk 5 above)
Comment (ucover) [reply frame/benchmarking/src/lib.rs:841]: I wondered the same, but thought that you probably only output the error stream.  
Resolving for now.

File: frame/benchmarking/src/lib.rs (hunk 2 above)
Comment (ucover) [reply frame/benchmarking/src/lib.rs:508]: Ok I see what you mean, it can also be understood that way. Fixed it (minus the `s` at the `generates`).

File: frame/benchmarking/src/lib.rs (hunk 3 above)
Comment (ucover) [reply frame/benchmarking/src/lib.rs:193]: The `skip_meta` was not obvious to me at all, doc would be much appreciated.

File: frame/democracy/src/benchmarking.rs [hunk 7]
```diff
@@ -70,7 +70,7 @@ fn add_referendum<T: Config>(n: u32) -> Result<ReferendumIndex, &'static str> {
 	let referendum_index: ReferendumIndex = ReferendumCount::<T>::get() - 1;
 	T::Scheduler::schedule_named(
 		(DEMOCRACY_ID, referendum_index).encode(),
-		DispatchTime::At(1u32.into()),
+		DispatchTime::At(2u32.into()),
```
Comment (thiolliere) [inline frame/democracy/src/benchmarking.rs:73]: why this change ?

File: frame/benchmarking/src/lib.rs (hunk 3 above)
Comment (thiolliere) [reply frame/benchmarking/src/lib.rs:193]: actually I don't know either for `skip_meta`, but we can document them in another PR, this PR just need to include the documentation for the new feature: i.e. document that we can write impl_benchmark_test_suite inside the benchmark macro directly in order to get a split test suite with potential better error message.

**Comment by thiolliere** [comment]
> > looks good to me once doc about new feature is added
> 
> So how does this work? Where do you want it documented? @thiolliere

I just mean this https://github.com/paritytech/substrate/pull/9860#discussion_r718897986 benchmark have a new feature: improved test suite when impl_benchmark_test_suite is written inside the macro. So that needs to be documented in the rust-doc of the item `benchmark`: here https://github.com/paritytech/substrate/blob/0d44d83e619f6c45597c11b37f3e36e942d22cc2/frame/benchmarking/src/lib.rs#L58-L191

File: frame/democracy/src/benchmarking.rs (hunk 7 above)
Comment (shawntabrizi) [reply frame/democracy/src/benchmarking.rs:73]: When running benchmarks, we set the block number to 1, versus before in tests, we had block number as 0.

This change avoids an error where we try to schedule an event to happen on the same block.

File: frame/benchmarking/src/lib.rs (hunk 3 above)
Comment (ucover) [reply frame/benchmarking/src/lib.rs:200]: Okay done @thiolliere

Comment (thiolliere) [reply frame/benchmarking/src/lib.rs:200]: thanks

**Review by thiolliere — APPROVED** [review]

**Comment by thiolliere** [comment]
I wonder if it would be difficult to do a test that ensures those test are correctly generated, maybe by ensuring they fail somewhere, anyway it is fine for now.

**Comment by thiolliere** [comment]
bot merge

**Comment by KiChjang** [comment]
Sorry I was late into the game, but I think a follow-up to this would be to figure out a nicer syntax for the `impl_benchmark_test_suite` macro. Since parsing and expanding the macro falls within the responsibility of `benchmarks!`, we don't actually need to keep macro syntax for `impl_benchmark_test_suite` anymore, and can think of a more intuitive one or even a less verbose one for it.

**Comment by thiolliere** [comment]
we could also make it an attribute macro so that people can have the automatic formatting

**Comment by ucover** [comment]
Nice bot! Thanks very much :smile:   

I added my polkadot address to the MR description.   
@shawntabrizi @substrate-tip-bot
